// @vue/language-core/lib/codegen/script/template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTemplate = generateTemplate;
exports.generateTemplateDirectives = generateTemplateDirectives;
const codeFeatures_1 = require("../codeFeatures");
const modules_1 = require("../style/modules");
const scopedClasses_1 = require("../style/scopedClasses");
const context_1 = require("../template/context");
const interpolation_1 = require("../template/interpolation");
const styleScopedClasses_1 = require("../template/styleScopedClasses");
const utils_1 = require("../utils");
const merge_1 = require("../utils/merge");
function* generateTemplate(options, ctx) {
    ctx.generatedTemplate = true;
    const templateCodegenCtx = (0, context_1.createTemplateCodegenContext)({
        scriptSetupBindingNames: new Set(),
    });
    yield* generateTemplateCtx(options);
    yield* generateTemplateElements();
    yield* generateTemplateComponents(options);
    yield* generateTemplateDirectives(options);
    yield* generateTemplateBody(options, templateCodegenCtx);
    return templateCodegenCtx;
}
function* generateTemplateCtx(options) {
    const exps = [];
    exps.push(`{} as InstanceType<__VLS_PickNotAny<typeof __VLS_self, new () => {}>>`);
    if (options.vueCompilerOptions.petiteVueExtensions.some(ext => options.fileName.endsWith(ext))) {
        exps.push(`globalThis`);
    }
    if (options.sfc.styles.some(style => style.module)) {
        exps.push(`{} as __VLS_StyleModules`);
    }
    yield `const __VLS_ctx = `;
    yield* (0, merge_1.generateSpreadMerge)(exps);
    yield utils_1.endOfLine;
}
function* generateTemplateElements() {
    yield `let __VLS_elements!: __VLS_IntrinsicElements${utils_1.endOfLine}`;
}
function* generateTemplateComponents(options) {
    const types = [`typeof __VLS_ctx`];
    if (options.sfc.script && options.scriptRanges?.exportDefault?.componentsOption) {
        const { componentsOption } = options.scriptRanges.exportDefault;
        yield `const __VLS_componentsOption = `;
        yield (0, utils_1.generateSfcBlockSection)(options.sfc.script, componentsOption.start, componentsOption.end, codeFeatures_1.codeFeatures.navigation);
        yield utils_1.endOfLine;
        types.push(`typeof __VLS_componentsOption`);
    }
    yield `type __VLS_LocalComponents = ${types.join(` & `)}${utils_1.endOfLine}`;
    yield `let __VLS_components!: __VLS_LocalComponents & __VLS_GlobalComponents${utils_1.endOfLine}`;
}
function* generateTemplateDirectives(options) {
    const types = [`typeof __VLS_ctx`];
    if (options.sfc.script && options.scriptRanges?.exportDefault?.directivesOption) {
        const { directivesOption } = options.scriptRanges.exportDefault;
        yield `const __VLS_directivesOption = `;
        yield (0, utils_1.generateSfcBlockSection)(options.sfc.script, directivesOption.start, directivesOption.end, codeFeatures_1.codeFeatures.navigation);
        yield utils_1.endOfLine;
        types.push(`__VLS_ResolveDirectives<typeof __VLS_directivesOption>`);
    }
    yield `type __VLS_LocalDirectives = ${types.join(` & `)}${utils_1.endOfLine}`;
    yield `let __VLS_directives!: __VLS_LocalDirectives & __VLS_GlobalDirectives${utils_1.endOfLine}`;
}
function* generateTemplateBody(options, templateCodegenCtx) {
    yield* (0, scopedClasses_1.generateStyleScopedClasses)(options, templateCodegenCtx);
    yield* (0, styleScopedClasses_1.generateStyleScopedClassReferences)(templateCodegenCtx, true);
    yield* (0, modules_1.generateStyleModules)(options);
    yield* generateCssVars(options, templateCodegenCtx);
    if (options.templateCodegen) {
        yield* options.templateCodegen.codes;
    }
    else {
        if (!options.scriptSetupRanges?.defineSlots) {
            yield `type __VLS_Slots = {}${utils_1.endOfLine}`;
        }
        yield `type __VLS_InheritedAttrs = {}${utils_1.endOfLine}`;
        yield `type __VLS_TemplateRefs = {}${utils_1.endOfLine}`;
        yield `type __VLS_RootEl = any${utils_1.endOfLine}`;
    }
}
function* generateCssVars(options, ctx) {
    if (!options.sfc.styles.length) {
        return;
    }
    yield `// CSS variable injection ${utils_1.newLine}`;
    for (const style of options.sfc.styles) {
        for (const binding of style.bindings) {
            yield* (0, interpolation_1.generateInterpolation)(options, ctx, style.name, codeFeatures_1.codeFeatures.all, binding.text, binding.offset);
            yield utils_1.endOfLine;
        }
    }
    yield `// CSS variable injection end ${utils_1.newLine}`;
}
